_books/ion-1-1/src/macros/defining_macros.md

```diff
@@ -56,17 +56,68 @@ To specify an encoding, the [parameter name](#parameter-names) is annotated with
 
 A parameter name may optionally be followed by a _cardinality modifier_. This is a sigil that indicates how many values the parameter expects the corresponding argument expression to produce when it is evaluated.
 
-| Modifier | Cardinality         |
-|:--------:|---------------------|
-|   `!`    | exactly-one value   |
-|   `?`    | zero-or-one value   |
-|   `+`    | one-or-more values  |
-|   `*`    | zero-or-more values |
+| Modifier |         Cardinality | Accepts single expression | Accepts empty group | Accepts populated group |
+|:--------:|--------------------:|:-------------------------:|:-------------------:|:-----------------------:|
+|   `?`    |   zero-or-one value |             ✅             |          ✅          |            ❌            |
+|   `*`    | zero-or-more values |             ✅             |          ✅          |            ✅            |
+|   `!`    |   exactly-one value |             ✅             |          ❌          |            ❌            |
+|   `+`    |  one-or-more values |             ✅             |          ❌          |            ✅            |
```

Comment:
I would not include this extension to the table. I understand what you're trying to get at, but I think it's best to keep the abstract model of macro arguments distinct from the encoding concerns for macro arguments.

My mental model of "Cardinality" is that it applies to the variable expansion. I.e. this is the essence of cardinality:
>If an argument expression expands to a number of values that the cardinality forbids, the reader must raise an error.


Then, restrictions on the encoding can be derived from the cardinality, and any such restrictions should be specified in the encoding sections (which you have done for example on L434 of `e_expressions.md`).

What do you think? Am I understanding things correctly?

Reply:
> I think it's best to keep the abstract model of macro arguments distinct from the encoding concerns for macro arguments. My mental model of "Cardinality" is that it applies to the variable expansion.

Unfortunately, cardinality is a confluence of encoding and expansion. I wish that it weren't--my preference was that it only affect encoding (`*`/`!` only) and that we'd denote expansion constraints separately (` -> string+`).  As it stands, the text encoding prohibits passing a populated expression group to `?` expressions. It also prohibits passing an empty expression group to ` +`. (Groups also cannot be passed to `!`, but that's not controversial.)

We need the text and binary encodings to be consistent enough in this regard that we can easily transcribe between the two. Either binary needs to prohibit the same combinations, or we allow variadics to use any encoding and only validate the expansion.

I'm open to removing this table edition for readability. However I think we should be _able_ to discuss them at this level because they're consistent across the encodings. Ideally, reproducing the binary section's table in the text section would be redundant.

Let's talk about this more later today.

Reply:
I've [removed the extension](https://github.com/amazon-ion/ion-docs/pull/348/commits/45084a9685848198c5205a0878dc35fa2078deda) in the interest of getting the rest of this merged. We can revisit the table later.